### PR TITLE
Expose is_signed for dtype

### LINKED ISSUE #29475: Expose a `dtype.is_signed`
## 🚀 Feature

I think it would be helpful to expose a `dtype.is_signed`, similar to dtype.is_floating_point. e.g. `torch.uint8.is_signed==False`.

## Motivation

would have been helpful when writing tests for:
https://github.com/pytorch/pytorch/pull/29231

I think we could update the tests once adding this.

### PR DESCRIPTION
**Pull request by vishwakftw**
Changelog:
- Expose is_signed for torch.dtype by modifying torch/csrc/Dtype.cpp
- Allow half, bfloat16 and bool to also been "known" by the isSignedType function

Test Plan:
- Add tests in test/test_torch.py

Closes #29475

### CODE REVIEW

File: c10/core/ScalarType.h [hunk 1]
```diff
@@ -317,7 +317,9 @@ static inline bool isSignedType(ScalarType t) {
       return std::numeric_limits<ctype>::is_signed;
 
     switch (t) {
-      AT_FORALL_SCALAR_TYPES_AND(Half, CASE_SIGNED)
+      AT_FORALL_SCALAR_TYPES_AND2(Half, Bool, CASE_SIGNED)
+      case ScalarType::BFloat16:
```
Comment (albanD) [inline c10/core/ScalarType.h]: Why is this one a special case?
Aren't there other special cases?

Comment (vishwakftw) [reply c10/core/ScalarType.h]: Oh, I seemed to have missed quantized dtypes. Thank you for pointing this out.

Comment (albanD) [reply c10/core/ScalarType.h]: There is a `toUnderlying ` function just above that might help with handling quantized dtypes?

Comment (vishwakftw) [reply c10/core/ScalarType.h]: I've fixed it by manually adding them, because I felt that using the macro would probably not reduce the size of the diff.

Comment (albanD) [reply c10/core/ScalarType.h]: Wouldn't adding `t = toUnderlying(t);` just before the switch work?

Comment (vishwakftw) [reply c10/core/ScalarType.h]: Oh yes, why didn’t I think of that. Sorry, let me fix this.

Comment (vishwakftw) [reply c10/core/ScalarType.h]: @albanD I've simplified it. Thanks for the review.

**Review by albanD — APPROVED** [review]
Thanks making the changes!

**Review by nairbv — APPROVED** [review]
Thanks!

**Comment by vishwakftw** [comment]
@albanD is this good to be merged?

**Comment by albanD** [comment]
@vbsinha Sorry I was checking with the people working on BFloat16 to make sure this special case was useful.
Given [this](https://github.com/pytorch/pytorch/blob/88e4cee3e70aac95dd2c18b898808ce3426cb3c9/c10/util/BFloat16-inl.h#L202-L217) I think you don't need this special case, and you can add it back to the main macro :)

**Comment by vishwakftw** [comment]
Sure, thanks @albanD.

Just a side note, I think you might have tagged the wrong person.

**Comment by albanD** [comment]
Yes, my bad, fixed that.
This will be good to merge once you do the change.

**Comment by vishwakftw** [comment]
@albanD I've fixed it. Thanks for the pointer about BFloat16.

**Comment by vishwakftw** [comment]
@albanD is this good to merge?

**Comment by albanD** [comment]
Yes, I was waiting for all the tests to run, failures are unrelated.

**Comment by gchanan** [comment]
Umm...are quantized dtypes signed?   It's not clear to me that the underlying dtype is relevant here, that's just a representation, i.e. I believe you can have negative values with a quint8 dtype.  Is that correct @jerryzh168 ?

**Comment by jerryzh168** [comment]
> Umm...are quantized dtypes signed? It's not clear to me that the underlying dtype is relevant here, that's just a representation, i.e. I believe you can have negative values with a quint8 dtype. Is that correct @jerryzh168 ?

yeah, quint8 dtype can represent negative floating point values, since we have scale and zero_point as well.

**Comment by jerryzh168** [comment]
We need to know how `is_signed` is used first, then we can decide what should be `is_signed` for quantized dtypes

**Comment by nairbv** [comment]
> We need to know how `is_signed` is used first, then we can decide what should be `is_signed` for quantized dtypes

Should we change this to throw an error for quantized types until we figure it out? The idea is for this to be usable from python, so could be used for anything. Within PyTorch I think so-far it's only used in this unit test:
https://github.com/pytorch/pytorch/pull/29231/files#diff-9996665f82f52030836eb8657057cfadR2645

**Comment by jerryzh168** [comment]
@nairbv yeah looks like `is_signed` doesn't make sense for quantized dtype in this case, we should throw an error if people call `is_signed` for quantized dtypes.

**Comment by nairbv** [comment]
@gchanan how do we feel about having access to a property raise an error? The behavior of torch.qint8.is_signed raising an error seems a bit odd to me where there's no function call, though should be an easy change to make.

**Comment by nairbv** [comment]
> though should be an easy change to make.
Actually it seems if you throw an error during a property access it dumps core.

**Comment by nairbv** [comment]
Nevermind, got it working:

https://github.com/pytorch/pytorch/pull/30527

Still seems a bit weird to me though to raise an error on an attribute, without a function call.

**Pull request closed without merging**